Refactor macro-defined LoadXXXCheckpointer fns to member template fns

RChrHill opened the pull request:
Replaces the macro-definitions of load checkpointer functions with wrappers around a generic template.
This template can be used to interface the HMCResourceManager with checkpointers that are not defined in Grid, which would remove the need for a divergent fork in order to test alternative serial formats.

RChrHill committed:
Refactor macro-defined LoadXXXCheckpointer fns to member template fns

edbennett commented:
Thanks Ryan, this seems to work well!

A few minor issues:
- One thing that I needed to do differently to the implementations in `HMCResourceManager` (presumably because I'm calling a method from outside the class) is to use `TheHMC.Resources.template LoadCheckpointer` rather than just `TheHMC.Resources.LoadCheckpointer`; this may be worth documenting somewhere.
- Can you please remove trailing whitespace?
- We discussed adding a comment in `CheckPointerModule` that we expect a public attribute `Name` (in the absence of being able to require this using C++20 syntax); could you add this please?
- More generally, do you think there is room to comment and otherwise document the new functionality more, so that the next person to to look at this has an easier time working out what's going on?

Thanks again!

edbennett committed:
remove trailing whitespace

edbennett committed:
add comment mentioning new interface requirement of CheckPointerModule

edbennett committed:
add additional documentation for new checkpointing functionality

aim to make life easier for anyone looking to implement their own checkpointers in future.

edbennett commented:
@RChrHill As discussed on Thursday, I've pushed some changes addressing my feedback above. No functional changes, only comments are added. If you think they are accurate, and improve the readability, then I'll merge this. If anything needs changing, please let me know!

RChrHill commented:
@edbennett This all looks good to me. Thanks for wrapping this up for me.

edbennett commented:
@RChrHill Thanks! Do you want to PR this upstream?
